lib/aliquot-pay.rb

```diff
@@ -89,23 +98,48 @@ def encrypt(cleartext_message)
 
   def build_payment_method_details
     return @payment_method_details if @payment_method_details
+
+    return build_payment_method_details_non_tokenized if @type == :browser
+
+    build_payment_method_details_tokenized
+  end
+
+  def build_payment_method_details_non_tokenized
     value = {
       'pan'             => @pan              || '4111111111111111',
       'expirationYear'  => @expiration_year  || Time.now.year + 1,
       'expirationMonth' => @expiration_month || 12,
-      'authMethod'      => @auth_method      || 'PAN_ONLY',
     }
 
-    if @auth_method == 'CRYPTOGRAM_3DS'
+    if @protocol_version == :ECv2 && @type == :browser
       value.merge!(
-        'cryptogram'   => @cryptogram    || 'SOME CRYPTOGRAM',
-        'eciIndicator' => @eci_indicator || '05'
+        'authMethod' => @auth_method || 'PAN_ONLY'
       )
     end
-
     value
   end
 
+  def build_payment_method_details_tokenized
+    value = {
+      'expirationYear'  => @expiration_year  || Time.now.year + 1,
+      'expirationMonth' => @expiration_month || 12,
+      'eciIndicator'  => @eci_indicator || '05'
+    }
+    if @protocol_version == :ECv1
+        value.merge!(
+          'dpan'          => @pan || '4111111111111111',
+          'authMethod'    => @auth_method || '3DS',
+          '3dsCryptogram' => @cryptogram    || Base64.strict_encode64(OpenSSL::Random.random_bytes(20))
+          )
+      else
+        value.merge!(
+          'pan'          => @pan || '4111111111111111',
+          'authMethod'    => @auth_method || 'CRYPTOGRAM_3DS',
+          'cryptogram'   => @cryptogram    || Base64.strict_encode64(OpenSSL::Random.random_bytes(20))
```

Comment:
```suggestion
      'expirationYear'  => @expiration_year  || Time.now.year + 1,
      'expirationMonth' => @expiration_month || 12,
      'eciIndicator'    => @eci_indicator    || '05'
    }
    if @protocol_version == :ECv1
        value.merge!(
          'dpan'          => @pan         || '4111111111111111',
          'authMethod'    => @auth_method || '3DS',
          '3dsCryptogram' => @cryptogram  || Base64.strict_encode64(OpenSSL::Random.random_bytes(20))
          )
      else
        value.merge!(
          'pan'        => @pan         || '4111111111111111',
          'authMethod' => @auth_method || 'CRYPTOGRAM_3DS',
          'cryptogram' => @cryptogram  || Base64.strict_encode64(OpenSSL::Random.random_bytes(20))
```

Reply:
Alternatively:

```suggestion
      'expirationYear'  => @expiration_year || Time.now.year + 1,
      'expirationMonth' => @expiration_month || 12,
      'eciIndicator'    => @eci_indicator || '05'
    }
    if @protocol_version == :ECv1
        value.merge!(
          'dpan'          => @pan || '4111111111111111',
          'authMethod'    => @auth_method || '3DS',
          '3dsCryptogram' => @cryptogram || Base64.strict_encode64(OpenSSL::Random.random_bytes(20))
          )
      else
        value.merge!(
          'pan'        => @pan || '4111111111111111',
          'authMethod' => @auth_method || 'CRYPTOGRAM_3DS',
          'cryptogram' => @cryptogram || Base64.strict_encode64(OpenSSL::Random.random_bytes(20))
```

Reply:
I prefer the first one

Reply:
Me too :slightly_smiling_face: 

It's in line with other parts of the file.